pkg/controller/machinepool/machinepool_controller.go

```diff
@@ -563,10 +609,18 @@ func (r *ReconcileMachinePool) syncMachineSets(
 	for i, ms := range generatedMachineSets {
 		found := false
 		for _, rMS := range remoteMachineSets.Items {
-			if matchAndMutate(pool, cd, ms, rMS) {
-				found = true
+			var err error
+			found, err = matchMachineSets(ms, rMS, logger)
+			if err != nil {
+				logger.WithError(err).Errorf("unable to match MachineSets %s and %s", ms.Name, rMS.Name)
+				continue // FIXME is this the correct behavior?
```

Comment:
Great question. Not all errors are created equal.

If we fail to match a rMS that we _should_ have matched, we'll end up creating a dup, and that's Bad™. In that case we would ideally want to fail the whole mpool reconcile.

BUT

If we fail the mpool due to an error parsing an _irrelevant_ rMS, it's kind of a DoS. (IOW a bad remote mset can bring down the whole mpool controller.) Ideally we would ignore that one and `continue`.

Like, ugh, but how about we figure a way to distinguish "labels matched, and the error happened after that". We can be fairly confident that label matching won't error (missing gMS label scenario notwithstanding), and it's reasonable to use that as a criterion for failing _this_ mpool reconcile.

So, let's give `matchMachineSets` a third return, `labelsMatched bool`, and use that to decide whether to error (`true`) or `continue` (`false`). With a nice code comment explaining the rationale.